game.cc

```diff
@@ -1,49 +1,606 @@
+				for(int y = 0; y < 8; y++){
+					//If it can find a valid move for the current player, returns false since it is not a stalemate
+					if (isValidMove(playerWhite[i]->getX(),playerWhite[i]->getY(),x,y) == true) return false;
+				}
+			}
+		}
+	}
+	else{
+		for (int i = 0; i < 16; i++){
+			if (playerBlack[i] == NULL) continue;
+			for(int x = 0; x < 8; x++){
+				for(int y = 0; y < 8; y++){
+					if (isValidMove(playerBlack[i]->getX(),playerBlack[i]->getY(),x,y) == true) return false;
+				}
+			}
+		}
+	}
+	return true;
+}
+
+bool Game::hasWon(){
+	return isCheckmate();
 }
 
 bool Game::isCheck(int player){
-	// placeholder
+	Piece* king;
+	if (player == WHITE){
+		//Finds the Location the the King the the array of pieces and create pointer to it
+		for (int i = 0; i < 25; i++){
+			if (playerWhite[i] != NULL && playerWhite[i]->getWorth() == KING){
+				king = playerWhite[i];
+				break;
+			}
+		}
+
+		//Goes through each opposition player and sees if any of them can make a valid move to the King
+		for (int i = 0; i < 25; i++){
+			if (playerBlack[i] != NULL){
+				if (playerBlack[i]->isMoveValid(king->getX(),king->getY()) == true){
+					return true;
+				}
+			}
+		}
+	}
+	else{
+		for (int i = 0; i < 25; i++){
+			if (playerBlack[i] != NULL && playerBlack[i]->getWorth() == KING){
+				king = playerBlack[i];
+				break;
+			}
+		}
+
+		for (int i = 0; i < 25; i++){
+			if (playerWhite[i] != NULL){
+				if (playerWhite[i]->isMoveValid(king->getX(),king->getY())){
+					return true;
+				}
+			}
+		}
+	}
 	return false;
 }
 
-bool Game::makeMove(int startX, int startY, int endX, int endY, char pawnPromotionPiece){
-	return false;
+bool Game::isCheck(){
+	isCheck(currentPlayer);
+}
+bool Game::makeMove(int startX, int startY, int endX, int endY, char promoteType, bool checkForCheck){
+	if (isValidMove(startX,startY,endX,endY) == false) return false;
+
+	//Checks to see if piece that it is moving to its own piece
+	for (int i = 0; i < 25; i++){
+		if (currentPlayer == WHITE){
+			if(playerWhite[i] == theBoard[endX][endY]) return false;
+		}
+		else{
+			if(playerBlack[i] == theBoard[endX][endY]) return false;
+		}
+	}
+
+	//Check for Check
+	if (checkForCheck == true){
+		if (isCheckAfterMove(startX,startY,endX,endY) == true) return false;
+	}
+
+	//Checks for Pawn Promotion,
+	if (promoteType != ' '){
+
+		//Checks that the piece is a pawn moving to the end row
+		if (theBoard[startX][startY]->getWorth() != PAWN) return false;
+		if (currentPlayer == WHITE && endY != 0) return false;
+		if (currentPlayer == BLACK && endY != 7) return false;
+
+
+
+		// Check if character is correct Case
+		if (promoteType > 'A') promoteType = promoteType - 'A' + 'a';
+
+		//Finds location of piece in the array of pieces
+		int loc = 16;
+		if (currentPlayer == WHITE){	
+			while(playerWhite[loc] != NULL){
+				++loc;
+			}
+		}
+		else{
+			while(playerBlack[loc] != NULL){
+				++loc;
+			}
+		}
+
+		if (promoteType == 'q'){
```

Comment:
You could shorten this code by writing:

Piece tmp = new Queen(currentPlayer);
tmp.setGame(this);
tmp.setLocation(startX,startY);
if (currentPlayer == WHITE) {
    playerWhite[loc] = tmp;
    theBoard[startX][startY] = playerWhite[loc];
} else {
    playerBlack[loc] = tmp;
    theBoard[startX][startY] = playerBlack[loc];
}
 tmp = 0;